File: CPP/ByTier/E/0792_E_Colored_Balls.cpp
// Problem: CF 792 E - Colored Balls
// https://codeforces.com/contest/792/problem/E

/*
 * Problem: E. Colored Balls
 * 
 * Purpose: 
 *   To divide all balls into minimum number of sets such that:
 *   - Each ball belongs to exactly one set.
 *   - No set is empty.
 *   - Each set contains only balls of one color.
 *   - The difference in sizes of any two sets is at most 1.
 * 
 * Approach:
 *   - Try different possible sizes for the sets.
 *   - For each possible size, check if it's valid using the 'check' function.
 *   - The valid size is then used to compute the minimum number of sets.
 * 
 * Time Complexity: O(sqrt(max_a) * n)
 * Space Complexity: O(1)
 * 
 * Algorithms/Techniques:
 *   - Binary search or brute-force on possible set sizes.
 *   - Greedy assignment of balls to sets.
 */

#include <stdio.h>
#include <algorithm>
#include <cmath>

using namespace std;
using ll = long long;
const int N = 1e6 + 10;
int a[N];
int tmp;
bool check(int sz, int rem, int n) {
  for (int i = 1; i <= n; ++i) {
    int x = a[i] / sz, y = a[i] % sz;
    if (y > x && rem)  // If remainder is greater than quotient and we have extra capacity, return false
      return 0;
    if (!rem) {
      if (y > x) {
        --sz;  // Reduce size of set to accommodate the remainder
        ++rem;  // Increase the remainder count
        x = a[i] / sz;
        y = a[i] % sz;
      }
      if (y > x)  // Still not feasible after adjustment
        return 0;
    }
  }
  tmp = sz;
  return 1;
}
int main() {
  int n;
  scanf("%d", &n);
  for (int i = 1; i <= n; ++i)
    scanf("%d", &a[i]);
  sort(a + 1, a + n + 1);  // Sort the array to simplify processing
  
  const int M = (int)sqrt(a[1]) + 1;  // Try divisors up to sqrt of the largest number
  for (int i = 1; i <= M; ++i) {
    int sz = a[1] / i, rem = a[1] % i;
    if (check(sz, rem, n))  // Check if this size works for all boxes
      break;
  }
  ll rs = 0;
  for (int i = 1; i <= n; ++i)
    rs += (a[i] + tmp) / (tmp + 1);  // Calculate number of sets for each type of ball
  printf("%lld\n", rs);
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/0792/0792_E_Colored_Balls.py
# Problem: CF 792 E - Colored Balls
# https://codeforces.com/contest/792/problem/E

"""
Algorithm: Minimize the number of sets to divide colored balls such that:
- Each set contains only balls of one color.
- The difference in size between any two sets is at most 1.
- All balls must be placed in sets.

Approach:
1. For each color, we have `a_i` balls of that color.
2. We are trying to find the optimal set size `sz` such that:
   - All balls of a color can be distributed into sets of size `sz` or `sz+1`.
   - The number of sets needed for all colors is minimized.
3. Try different possible values of `sz` (from the square root of the minimum number of balls).
   - For each such `sz`, check if that `sz` works (i.e., all `a_i` can be split into groups of size `sz` or `sz+1`).
4. The number of sets required for a color `a_i` is `ceil(a_i / (sz + 1))`.

Time Complexity: O(sqrt(min(a_i)) * n)
Space Complexity: O(1)

This solution optimizes by checking divisors of the minimum number of balls and related values.
"""

import math

n = int(input())
s = input().split(" ")
for i in range(len(s)):
    s[i] = int(s[i])
m = min(s)  # Find the minimum number of balls of any color


def try_sz(x):
    # Check if x is a viable set size (all a_i can be distributed into sets of size x or x+1)
    for thing in s:
        if thing % x > math.floor(thing / x):
            return False
    return True


good = False
sz = 0
# Try divisors and related values around sqrt(m) for potential set size
for i in range(1, math.ceil(m**0.5) + 1):
    if try_sz(math.floor(m / i)):
        good = True
        sz = math.floor(m / i)
        break
    if m % i == 0:
        if try_sz(math.floor(m / i) - 1):
            good = True
            sz = math.floor(m / i) - 1
            break

ans = 0
# Compute total number of sets needed for all colors
for thing in s:
    ans += math.ceil(thing / (sz + 1))
print(str(ans))


# https://github.com/VaHiX/CodeForces/